File: Sorting/quickSort.py
"""
Merge sort:
Time Complexity - O(nlogn)
Space Complexity - O(1)
"""
import math
comparisons = 0


def quicksort(arr, left, right):
    global comparisons
    if left < right:
        comparisons += right - left
        pivotIndex = partition(arr, left, right)
        quicksort(arr, left, pivotIndex - 1)
        quicksort(arr, pivotIndex + 1, right)
        return arr


def partition(arr, left, right):

    ## Choosing median as the pivot element
    # pivotIndex = int(math.ceil((right+1+left)/2)) - 1

    ## Choosing first element as the pivot element
    pivotIndex = left

    ## Choosing last element as the pivot element
    # pivotIndex = right

    pivotValue = arr[pivotIndex]
    i = left

    for j in range(left, right + 1):
        if arr[j] < pivotValue:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1

    arr[arr.index(pivotValue)], arr[i] = arr[i], arr[arr.index(pivotValue)]

    return i



									### Test Cases ###
ls = [2,1,4,8,7,5,3,12,6,10]
print (ls)
print (quicksort(ls, 0, len(ls) - 1))
print (comparisons)

File: Sorting/quickSort.cpp
/*
Quick Sort 

Time Complexity - O(n^2) in the worst case; O(nlogn) when randomized partitioning is performed
Space Complexity - O(log n), since logn is negligible, quick sort can be claimed as In-place sorting algorithm
*/

#include <iostream>
using namespace std;

void swap(int * arr, int x, int y){
    int temp;
    temp = arr[x];
    arr[x] = arr[y];        
    arr[y] = temp;
}


int partition(int* arr, int start, int end){
    int pivotValue = arr[end];
    int pivotIndex = start, temp;

    for(int i = start; i < end; i++){
        if(arr[i] <= pivotValue){
            swap(arr, i, pivotIndex);
            pivotIndex++;
        }
    }

    swap(arr, pivotIndex, end);
    return pivotIndex;
}


void quickSort(int* arr, int start, int end){
    if(start < end){
        // O(n)
        int partitionIndex = partition(arr, start, end);
        quickSort(arr, start, partitionIndex - 1);
        quickSort(arr, partitionIndex + 1, end);
    }
}


int main(){
    int arr[] = {9, 183, 23 , 46, 45, 22, 12, 3};
    quickSort(arr, 0, sizeof(arr) / sizeof(int) - 1);

    for(int i = 0; i < sizeof(arr) / sizeof(int); i++)  cout<<arr[i] << " "; 
    
    return 0;
}